rapid/src/test/java/com/vrg/rapid/MessagingTest.java

```diff
@@ -123,145 +117,25 @@ public void joinFirstNodeRetryWithErrors()
         // Try with the same host details as the server
         final Endpoint clientAddr1 = Utils.hostFromParts(LOCALHOST_IP, serverPort);
         final GrpcClient client1 = new GrpcClient(clientAddr1);
-        final JoinResponse result1 = sendPreJoinMessage(client1, serverAddr, clientAddr1,
+        final JoinResponse result1 = sendJoinMessage(client1, serverAddr, clientAddr1,
                                                              Utils.nodeIdFromUUID(UUID.randomUUID()));
         assertNotNull(result1);
         assertEquals(JoinStatusCode.HOSTNAME_ALREADY_IN_RING, result1.getStatusCode());
-        assertEquals(K, result1.getEndpointsCount());
+        assertEquals(0, result1.getEndpointsCount());
         assertEquals(0, result1.getIdentifiersCount());
 
         // Try again with a different port, this should fail because we're using the same
         // uuid as the server.
         final int clientPort2 = 1235;
         final Endpoint clientAddr2 = Utils.hostFromParts(LOCALHOST_IP, clientPort2);
         final GrpcClient client2 = new GrpcClient(clientAddr2);
-        final JoinResponse result2 = sendPreJoinMessage(client2, serverAddr, clientAddr2, nodeIdentifier);
+        final JoinResponse result2 = sendJoinMessage(client2, serverAddr, clientAddr2, nodeIdentifier);
         assertNotNull(result2);
         assertEquals(JoinStatusCode.UUID_ALREADY_IN_RING, result2.getStatusCode());
         assertEquals(0, result2.getEndpointsCount());
         assertEquals(0, result2.getIdentifiersCount());
     }
 
-    /**
-     * A node in a cluster gets a join request from a peer with non conflicting
```

Comment:
These tests are being removed because they involve exercising the old "join phase 1" step. That said, it would be good to add new tests to make sure that some of the defenses produced by the two rounds still holds.

Some scenarios that come to mind -- what happens if:
1) a joiner contacts one node, gets admitted to the membership but does not learn about it (say, because of network issues), and can't reach the first node after. What are it's options? Does it contact a second node? Or does the join call fail as was the old behavior (and it retries with a new ID)?
2) a joiner contacts a node A while A has already issued a proposal and is preparing to move to a new configuration?

Reply:
1) If it can reach the node again, then the `SAME_NODE_ALREADY_IN_RING` behavior will allow it to join (if the attempts do not run out before it has a chance to get that response). If it can't contact the same node again, and the application using the library attempts to join another potential seed node, then that join attempt will fail (because the UUID will have changed). The node will have to be patient and wait until the failure detectors have kicked in and a new membership has been agreed upon. I think this behavior is actually quite okay. In Akka's cluster protocol, nodes in this situation would need to be removed manually - I mean, it'd be an option to offer some kind of admin interface that'd allow to manually remove a node, but I think this is a bit too much (especially since failure detection should only be taking a few seconds).

2) I have to think about this a bit more.

Agree that this should be tested for, hope I'll get around to it next week

Reply:
I'm fine with 1., as long as there are safeguards in place for the node to not attempt a join against two different nodes in the same round. It's best that the join call throws an exception and the user code retries.

Reply:
1) Should be fixed now

I also fixed 2) by checking if a view change is in progress and returning that, so that the joining node retries. I had to change `MembershipView` to be non-final in order to be able to simulate the concurrent behavior.